src/main/java/net/minecraftforge/common/util/ValidBlocksSet.java

```diff
@@ -0,0 +1,23 @@
+package net.minecraftforge.common.util;
+
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.GatherValidBlocksEvent;
+
+import java.util.Set;
+
+public final class ValidBlocksSet {
+    private final Set<Block> original;
+    private volatile Set<Block> computed = null;
+
+    public ValidBlocksSet(Set<Block> original) {
+        this.original = original;
+    }
+
+    public boolean contains(Block block, BlockEntityType<?> type) {
```

Comment:
As i said on discord, I think it would be better to fire this event at a known time, during post registration, instead of on first use. Theoretically it'd be fine as this method should only be used when the entries are loaded into world which is far after registration. But its a race and i'd rather just be clear on when its fired.